File: Parser/tokenizer.py
import re

from Mbase.error import print_error_with_origin
from Mbase.types import VALID_DIGITS
from Parser.token_type import TokenType
from Parser.token import Token

def tokenize(source):
    i = 0
    while i < len(source):
        char = source[i]

        # Base literal: b10@123
        if char == 'b':
            base_match = re.match(r"b(\d{1,2})@", source[i:])
            if base_match:
                base = int(base_match.group(1))
                j = i + base_match.end()

                if j < len(source) and source[j] == '(':  # b10@(...)
                    j += 1
                    start = j
                    while j < len(source) and source[j] != ')':
                        j += 1
                    if j == len(source):
                        print_error_with_origin(source, i, "Unclosed base literal", label="SyntaxError")
                        return
                    raw = source[start:j]
                    j += 1
                else:
                    start = j
                    while j < len(source) and source[j].lower() in VALID_DIGITS[:base]:
                        j += 1
                    raw = source[start:j]

                yield Token(TokenType.BASE_LITERAL, (base, raw), i)
                i = j
                continue

        # Skip comments
        if char == '#':
            while i < len(source) and source[i] != '\n':
                i += 1
            continue

        # Newline
        elif char.isspace():
            if char == '\n':
                yield Token(TokenType.NEWLINE, '\n', i)
            i += 1
            continue

        # Multi-character operators
        elif source[i:i+3] == '===':
            yield Token(TokenType.STRICT_EQUAL, '===', i)
            i += 3
        elif source[i:i+3] == '!==':
            yield Token(TokenType.STRICT_NOTEQUAL, '!==', i)
            i += 3
        elif source[i:i+2] == '==':
            yield Token(TokenType.EQUAL, '==', i)
            i += 2
        elif source[i:i+2] == '!=':
            yield Token(TokenType.NOTEQUAL, '!=', i)
            i += 2
        elif source[i:i+2] == '<=':
            yield Token(TokenType.LEQ, '<=', i)
            i += 2
        elif source[i:i+2] == '>=':
            yield Token(TokenType.GEQ, '>=', i)
            i += 2
        elif source[i:i+2] == '&&':
            yield Token(TokenType.AND, '&&', i)
            i += 2
        elif source[i:i+2] == '||':
            yield Token(TokenType.OR, '||', i)
            i += 2

        # Single-character symbols
        elif char == ';':
            yield Token(TokenType.SEMICOLON, ';', i); i += 1
        elif char == ',':
            yield Token(TokenType.COMMA, ',', i); i += 1
        elif char == '{':
            yield Token(TokenType.LBRACE, '{', i); i += 1
        elif char == '}':
            yield Token(TokenType.RBRACE, '}', i); i += 1
        elif char == '(':
            yield Token(TokenType.LPAREN, '(', i); i += 1
        elif char == ')':
            yield Token(TokenType.RPAREN, ')', i); i += 1
        elif char == '=':
            yield Token(TokenType.ASSIGN, '=', i); i += 1
        elif char == '+':
            yield Token(TokenType.PLUS, '+', i); i += 1
        elif char == '-':
            yield Token(TokenType.MINUS, '-', i); i += 1
        elif char == '*':
            yield Token(TokenType.STAR, '*', i); i += 1
        elif char == '/':
            yield Token(TokenType.SLASH, '/', i); i += 1
        elif char == '<':
            yield Token(TokenType.LESSTHAN, '<', i); i += 1
        elif char == '>':
            yield Token(TokenType.GREATERTHAN, '>', i); i += 1
        elif char == '!':
            yield Token(TokenType.NOT, '!', i); i += 1
        elif char == '@':
            yield Token(TokenType.AT, '@', i); i += 1

        # String literals
        elif char == '"':
            start = i
            i += 1
            content = ''
            while i < len(source):
                if source[i] == '\\' and i + 1 < len(source):
                    esc = source[i+1]
                    content += {'n':'\n','t':'\t','"':'"','\\':'\\'}.get(esc, esc)
                    i += 2
                elif source[i] == '"':
                    i += 1
                    break
                else:
                    content += source[i]
                    i += 1
            yield Token(TokenType.TEXT, content, start)

        # Numbers
        elif re.match(r'[0-9]', char):
            start = i
            while i < len(source) and source[i].isdigit():
                i += 1
            yield Token(TokenType.NUMBER, source[start:i], start)

        # Identifiers and keywords
        elif re.match(r'[a-zA-Z_]', char):
            start = i
            while i < len(source) and re.match(r'[a-zA-Z0-9_]', source[i]):
                i += 1
            value = source[start:i]
            type_map = {
                "fn": TokenType.FUNCTION,
                "ret": TokenType.RETURN,
                "if": TokenType.IF,
                "else": TokenType.ELSE,
                "while": TokenType.WHILE,
                "loop": TokenType.LOOP,
                "break": TokenType.BREAK,
                "continue": TokenType.CONTINUE,
            }
            tok_type = type_map.get(value, TokenType.IDENTIFIER)
            yield Token(tok_type, value, start)

        # Error for unknown chars
        else:
            print_error_with_origin(source, i, f"Unexpected character: {char}", label="SyntaxError")
            raise SyntaxError(f"Unexpected character: {char}")

    yield Token(TokenType.EOF, None, len(source))

File: rust/src/tokenizer.rs
use crate::token::{Token, TokenType};
use crate::base_literal::VALID_DIGITS;

pub fn tokenize(source: &str) -> Vec<Token> {
    let mut tokens = Vec::new();
    let mut i = 0;
    let chars: Vec<char> = source.chars().collect();
    while i < chars.len() {
        let ch = chars[i];
        if ch == 'b' {
            let mut j = i + 1;
            while j < chars.len() && chars[j].is_ascii_digit() { j += 1; }
            if j < chars.len() && chars[j] == '@' {
                let base: u32 = source[i+1..j].parse().unwrap();
                j += 1;
                let start = j;
                while j < chars.len() && VALID_DIGITS[..base as usize].contains(chars[j]) { j += 1; }
                let raw = &source[start..j];
                tokens.push(Token::new(TokenType::BaseLiteral, Some(format!("{}@{}", base, raw)), i));
                i = j;
                continue;
            }
        }
        if ch == '#' {
            while i < chars.len() && chars[i] != '\n' { i += 1; }
            continue;
        }
        if ch.is_whitespace() {
            if ch == '\n' { tokens.push(Token::new(TokenType::Newline, Some("\n".into()), i)); }
            i += 1;
            continue;
        }
        match ch {
            ';' => { tokens.push(Token::new(TokenType::Semicolon, Some(";".into()), i)); i += 1; }
            ',' => { tokens.push(Token::new(TokenType::Comma, Some(",".into()), i)); i += 1; }
            '{' => { tokens.push(Token::new(TokenType::LBrace, Some("{".into()), i)); i += 1; }
            '}' => { tokens.push(Token::new(TokenType::RBrace, Some("}".into()), i)); i += 1; }
            '(' => { tokens.push(Token::new(TokenType::LParen, Some("(".into()), i)); i += 1; }
            ')' => { tokens.push(Token::new(TokenType::RParen, Some(")".into()), i)); i += 1; }
            '=' => { tokens.push(Token::new(TokenType::Assign, Some("=".into()), i)); i += 1; }
            '+' => { tokens.push(Token::new(TokenType::Plus, Some("+".into()), i)); i += 1; }
            '-' => { tokens.push(Token::new(TokenType::Minus, Some("-".into()), i)); i += 1; }
            '*' => { tokens.push(Token::new(TokenType::Star, Some("*".into()), i)); i += 1; }
            '/' => { tokens.push(Token::new(TokenType::Slash, Some("/".into()), i)); i += 1; }
            '<' => { tokens.push(Token::new(TokenType::LessThan, Some("<".into()), i)); i += 1; }
            '>' => { tokens.push(Token::new(TokenType::GreaterThan, Some(">".into()), i)); i += 1; }
            '!' => { tokens.push(Token::new(TokenType::Not, Some("!".into()), i)); i += 1; }
            '@' => { tokens.push(Token::new(TokenType::At, Some("@".into()), i)); i += 1; }
            '"' => {
                let start = i+1;
                i += 1;
                let mut content = String::new();
                while i < chars.len() && chars[i] != '"' {
                    if chars[i] == '\\' && i + 1 < chars.len() {
                        let esc = chars[i+1];
                        let val = match esc { 'n' => '\n', 't' => '\t', '"' => '"', '\\' => '\\', _ => esc };
                        content.push(val);
                        i += 2;
                    } else {
                        content.push(chars[i]);
                        i += 1;
                    }
                }
                i += 1;
                tokens.push(Token::new(TokenType::Text, Some(content), start-1));
            }
            _ => {
                if ch.is_ascii_digit() {
                    let start = i; while i < chars.len() && chars[i].is_ascii_digit() { i += 1; }
                    tokens.push(Token::new(TokenType::Number, Some(source[start..i].into()), start));
                } else if ch.is_ascii_alphanumeric() || ch == '_' {
                    let start = i; while i < chars.len() && (chars[i].is_ascii_alphanumeric() || chars[i] == '_') { i += 1; }
                    let value = &source[start..i];
                    let ttype = match value {
                        "fn" => TokenType::Function,
                        "ret" => TokenType::Return,
                        "if" => TokenType::If,
                        "else" => TokenType::Else,
                        "while" => TokenType::While,
                        "loop" => TokenType::Loop,
                        "break" => TokenType::Break,
                        "continue" => TokenType::Continue,
                        _ => TokenType::Identifier,
                    };
                    tokens.push(Token::new(ttype, Some(value.into()), start));
                } else {
                    i += 1;
                }
            }
        }
    }
    tokens.push(Token::new(TokenType::Eof, None, source.len()));
    tokens
}
